run/fvde2john.py

```diff
@@ -0,0 +1,220 @@
+            print(f"[!] Error reading entry: {e}")
+
+def recover_file(fs_object, file_path):
+    try:
+        file_obj = fs_object.open(file_path)
+        size = file_obj.info.meta.size
+        offset = 0
+        data = file_obj.read_random(offset, size)
+
+        return data
+
+    except Exception as e:
+        print(f"[!] Error recovering file: {e}")
+
+def get_EncryptedRoot_plist_wipekey(image_file, start_pos):
+    img = pytsk3.Img_Info(image_file)
+    fs = pytsk3.FS_Info(img, offset=start_pos)
+    target_file = 'EncryptedRoot.plist.wipekey'
+    EncryptedRoot_data = traverse_filesystem(fs, target_file)
+
+    return EncryptedRoot_data
+
+def main():
+
+    p = argparse.ArgumentParser()
+    p.add_argument('image_file')
+    args = p.parse_args()
+    image_file = args.image_file
+
+    with open(image_file, 'rb') as fp:
+        partition_dict = parse_partition_table(fp)
+
+        core_storage_volumes = get_all_partitions_of_type(partition_dict, HEX_CORE_STORAGE_TYPE_GUID)
+        boot_volumes = get_all_partitions_of_type(partition_dict, HEX_APPLE_BOOT_STORAGE_TYPE_GUID)
+
+        # Unlikely to have more than one boot volume, but loop anyway
+        for boot_start_pos in boot_volumes:
+            EncryptedRoot_data = get_EncryptedRoot_plist_wipekey(image_file, boot_start_pos)
+
+            if core_storage_volumes == []:
+                print("[!] No CoreStorage volumes found, exiting.")
+                exit()
+            for cs_start_pos in core_storage_volumes:
+                aes_key = parse_corestorage_header(fp, cs_start_pos)
+
+                tweak = b'\x00' * 16
+                pt = AES_XTS_decrypt(aes_key, tweak, EncryptedRoot_data)
+                d = load_plist_dict(pt)
+
+                userIndex = 0
+                for i in range(len(d['CryptoUsers'])):
+                    # We want the local user login details i.e. not iCloud
+                    if d['CryptoUsers'][i].get('UserType') == LOCAL_USER_TYPE_ID:
+                        userIndex = i
+
+                        print("\n[+] Finding user info that may be useful to crack the password:")
+                        FullName = d['CryptoUsers'][userIndex].get('UserFullName')
+                        PassphraseHint = d['CryptoUsers'][userIndex].get('PassphraseHint')
+
+                        print("Full name:", FullName)
+                        print("Passphrase hint:", PassphraseHint)
+
+                        PassphraseWrappedKEKStruct = d['CryptoUsers'][userIndex].get('PassphraseWrappedKEKStruct')
+                        print("[+] Found PassphraseWrappedKEKStruct in decrypted plist")
+
+                        salt = PassphraseWrappedKEKStruct[8:24]
+                        kek  = PassphraseWrappedKEKStruct[32:56]
+                        iterations = uint_to_int(PassphraseWrappedKEKStruct[168:172])
+
+                        print(f"[+] Successfully extracted hash: $fvde$1${len(salt)}${salt.hex()}${iterations}${kek.hex()}")
```

Comment:
I'm a python n00b but to me this looks like it's literally printing something like
```
[+] Successfully extracted hash: $fvde$1$(...)
```
If so, maybe the "[+] Successfully extracted hash" should go to stderr, if kept at all?

Reply:
Yes, all messages in the script currently go to stdout. This is in contrast to most other scripts, where only the hash is printed to stdout and everything else to stderr.
I would have recommended to adapt to this behavior. Though the difference to other scripts is that this prints additional information like username and password hint.
But yes, at the moment it is untypically verbose.

Reply:
I think one of the script's output lines (on stdout) should be an /etc/passwd-format line without any sort of `[+] Successfully extracted hash: ` prefix or such. So that if output is redirected to a file, that file is directly and optimally usable with `john`, without any need to manually select a portion of output and copy-paste.